move StorageClass from beta to stable

jsafrane opened the pull request:
There is one PRs on review to get necessary API objects. When it is merged, there is still a lot of work to do. This PR is a preview of this work. I think that's all that needs to be done, I'll update docs and check examples afterwards.

Expected behavior:
- both beta and StorageClassName attribute are set: beta is used
- only beta is set: beta is used
- only attribute is set: attribute is used

Rest of the commits is "just code" with little invention.

@msau42 @saad-ali @kerneltime @jingxu97 @kubernetes/sig-storage-pr-reviews 

Release note:
```release-note
StorageClassName attribute has been added to PersistentVolume and PersistentVolumeClaim objects and should be used instead of annotation `volume.beta.kubernetes.io/storage-class`. The beta annotation is still working in this release, however it will be removed in a future release.
```

jsafrane committed:
Update bazel

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41646)
<!-- Reviewable:end -->

jsafrane commented:
marking as do-not-merge, waiting for #40088 and #40013 merged first

jsafrane commented:
rebased, I hope at least some test passes :-)

msau42 commented on pkg/kubectl/describe.go:
```diff
@@ -830,7 +830,7 @@ func (d *PersistentVolumeDescriber) Describe(namespace, name string, describerSe
 		w := &PrefixWriter{out}
 		w.Write(LEVEL_0, "Name:\t%s\n", pv.Name)
 		printLabelsMultiline(w, "Labels", pv.Labels)
-		w.Write(LEVEL_0, "StorageClass:\t%s\n", storageutil.GetStorageClassAnnotation(pv.ObjectMeta))
```
Would a newer kubectl ever have to talk to an older API server?

msau42 replied:
Likewise, would an older kubelet have to talk to a newer API server?

msau42 replied:
Ok I think I see in conversion.go, it will convert beta annotation to spec.storageclass.  So newer kubectl reading from old API will work.

And old kubectl sending beta annotation -> new API will work.  But what about old kubectl reading from new API that doesn't use beta annotation?

jsafrane replied:
it is easily possible to add an annotation when converting from internal object to (externa) V1 one, so old applications (kubectl) will work.

There is one glitch though, I would need to mess up with serialization unit test (and probably fuzz tests too) for PVs and PVCs. The tests serialize bunch of weird API objects and they expect to get the same object after de-serialization. There is also internal -> v1 conversion done somewhere in the process that would add the annotation and therefore the input and output objects would differ. The code is scattered among several files, see e.g. Pod.Spec.InitContainers, they convert v1 beta annotation -> internal attribute in the same way as this PR:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unstructured_test.go#L48
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/conversion_test.go#L38
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/serialization_proto_test.go#L65
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/serialization_test.go#L387

I don't like this hacking of tests at all. Is strict backward compatibility with old kubectl necessary?

msau42 replied:
I don't think it's necessary, since this would be an issue anytime someone adds a new field.  The main requirement is that older clients can create and view their older objects, which this handles.

msau42 commented on pkg/api/helpers.go:
```diff
@@ -565,3 +565,16 @@ func PodAnnotationsFromSysctls(sysctls []Sysctl) string {
+// when DefaultStorageClass admission plugin is disabled or there is no default
+// storage class configured - caller does not need to check for
+// storageClassName == nil.
+func GetClaimStorageClass(claim *PersistentVolumeClaim) string {
```
When do you use api.GetClaimStorageClass vs v1.GetClaimStorageClass?

jsafrane replied:
I think admission plugins and kubectl use internal api, everything else (incl. controllers) use v1.

jsafrane replied:
And I don't like GetClaimStorageClass defined twice either. Suggestions are welcome.

msau42 replied:
Can v1.GetClaimStorageClass just call api.GetClaimStorageClass?

jsafrane committed:
PV controller: use attributes instead of beta annotations in unit tests

jsafrane committed:
integration tests: use attributes instead of beta annotations

jsafrane committed:
quota: use attributes instead of beta annotations

jsafrane committed:
admission controller: use attributes instead beta annotations

jsafrane committed:
e2e tests: use attributes instead of beta annotations

jsafrane committed:
Add StorageClassName validation

jsafrane commented on pkg/controller/volume/persistentvolume/pv_controller.go:
```diff
@@ -1336,16 +1336,13 @@ func (ctrl *PersistentVolumeController) provisionClaimOperation(claimObj interfa
 	// Bind it to the claim
 	volume.Spec.ClaimRef = claimRef
 	volume.Status.Phase = v1.VolumeBound
+	if volume.Spec.StorageClassName == "" {
```
The code here makes sure that provisioned PV has the class set correctly - either by the provisioner (`provisioner.Provision()` call on line 1321), or if the provisioner forgets to set it, the controller does it here.

jsafrane commented on pkg/api/v1/conversion_test.go:
```diff
@@ -216,3 +217,199 @@ func TestResourceListConversion(t *testing.T) {
+		{ // Annotation is copied to class
+			input: claimTestSpec{
+				annotations: map[string]string{
+					storageutil.BetaStorageClassAnnotation: "foo",
```
Sure, good point.

jsafrane replied:
Fixed as separate commit `[squash] Use variables in conversion unit tests.`

jsafrane commented:
@k8s-bot verify test this

thockin commented:
In general we do not want to do automatic up-conversion.  It makes ambiguity that we don't want to deal with - specifically what happens if we find both things set?  Better to just be explicit. I think.

jsafrane commented:
> In general we do not want to do automatic up-conversion. It makes ambiguity that we don't want to deal with - specifically what happens if we find both things set? Better to just be explicit. I think.

I hate reinventing the wheel so I'm following Pod.Spec.InitContainers, which went from beta annotations to attribute in 1.6 and they do very similar conversion as I do: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/conversion.go#L402. In case both the annotation and the attribute are set, InitContainer beta annotation wins (!). Looking at git blame, it comes from @smarterclayton.

I admit, in my PR the class attribute wins, which is inconsistent with InitContainers (and I can fix it), however the precedent is already here.

thockin commented:
The beta has to win, to allow older clients that don't know about the
field, but it is still ambiguous.

It means that "new" clients have to edit both the field and annotation.
You can never just update the field as long as both exist.  My position is
that it is more obvious just to require the user to own the conversion.

On Feb 23, 2017 4:20 AM, "Jan Šafránek" <notifications@github.com> wrote:

> In general we do not want to do automatic up-conversion. It makes
> ambiguity that we don't want to deal with - specifically what happens if we
> find both things set? Better to just be explicit. I think.
>
> I hate reinventing the wheel so I'm following Pod.Spec.InitContainers,
> which went from beta annotations to attribute in 1.6 and they do very
> similar conversion as I do: https://github.com/kubernetes/
> kubernetes/blob/master/pkg/api/v1/conversion.go#L402. In case both the
> annotation and the attribute are set, InitContainer beta annotation wins
> (!). Looking at git blame, it comes from @smarterclayton
> <https://github.com/smarterclayton>.
>
> I admit, in my PR the class attribute wins, which is inconsistent with
> InitContainers (and I can fix it), however the precedent is already here.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41646#issuecomment-281977616>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVFEdLOCSs2dpVwxsa74P6O618uKLks5rfXmXgaJpZM4MEczh>
> .
>

msau42 commented:
Tim, are you suggesting that v1 clients should not be allowed to use the beta annotation?  That way, if we see a beta annotation, then we know it must be coming from an older client, and we don't need to handle the case where both are set?

jsafrane commented:
Rebased, squashed and updated "Add StorageClassName conversion logic" patch:

> The beta has to win, to allow older clients that don't know about the field

Fixed. Current conversion behavior:

* nothing is set -> we're waiting for admission plugin to set default
* beta annotation is set, attribute is nil -> beta is copied to the attribute
* **beta annotation is set, attribute is set -> beta is copied to the attribute**
* beta is not set, attribute is set -> nothing happens

Just a reminder, rest of the code then uses the attribute for all decisions and does not look at annotation.

All other patches should be the same as before.

thockin commented:
It's the "if both are set, blow away the user's value" that is wrong.  We
shouldn't be changing user data.  This is why the up-conversion is
problematic.

On Fri, Feb 24, 2017 at 6:52 AM, Jan Šafránek <notifications@github.com>
wrote:

> Rebased, squashed and updated "Add StorageClassName conversion logic"
> patch:
>
> The beta has to win, to allow older clients that don't know about the field
>
> Fixed. Current conversion behavior:
>
>    - nothing is set -> we're waiting for admission plugin to set default
>    - beta annotation is set, attribute is nil -> beta is copied to the
>    attribute
>    - *beta annotation is set, attribute is set -> beta is copied to the
>    attribute*
>    - beta is not set, attribute is set -> nothing happens
>
> Just a reminder, rest of the code then uses the attribute for all
> decisions and does not look at annotation.
>
> All other patches should be the same as before.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41646#issuecomment-282310064>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVCEaHUpEqoGQYSd2bVyV7f3LeAQIks5rfu6VgaJpZM4MEczh>
> .
>

msau42 commented:
So then in the case where both the annotation and the field are set, the validation should throw an error.  The only cases that can't be handled are the default storageclasses.  For PV, we can't distinguish between spec.StorageClass == "" and unset.  And for PVC, we cannot distinguish between spec.StorageClass == nil and unset.  In those two cases, we'll still have to take the beta annotation.

thockin commented:
I don't think it needs to be an error.  There's no real precedent for
that.  The case of both being set is an edge-case - we should just do the
simplest things possible.  I totally ACK the desire to do something more
consistent, but there are so many weird cases to consider wrt compat.

On Fri, Feb 24, 2017 at 2:32 PM, Michelle Au <notifications@github.com>
wrote:

> So then in the case where both the annotation and the field are set, the
> validation should throw an error. The only cases that can't be handled are
> the default storageclasses. For PV, we can't distinguish between
> spec.StorageClass == "" and unset. And for PVC, we cannot distinguish
> between spec.StorageClass == nil and unset. In those two cases, we'll still
> have to take the beta annotation.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41646#issuecomment-282423647>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVHZ4yRmBDvO8x5gyhJDhVGj5VQ6Mks5rf1pkgaJpZM4MEczh>
> .
>

msau42 commented:
Ok, so if I understand correctly, overriding the user's spec value with the beta value is bad.  Perhaps instead, we could use a wrapper function to get the storageclass name, that will handle the beta and v1 cases.

thockin commented:
a helper function is a pretty common pattern :)

On Fri, Feb 24, 2017 at 3:43 PM, Michelle Au <notifications@github.com>
wrote:

> Ok, so if I understand correctly, overriding the user's spec value with
> the beta value is bad. Perhaps instead, we could use a wrapper function to
> get the storageclass name, that will handle the beta and v1 cases.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41646#issuecomment-282435629>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVEmxA5e4dGza8khE2tk4RdidmzEJks5rf2slgaJpZM4MEczh>
> .
>

jsafrane committed:
PV controller: Set both beta annotation and StorageClassName during provisioning

jsafrane commented:
Rebased & reworked to helper functions instead of conversion magic (see the first commit). Beta annotation still has priority over `PV/PVC.Spec.StorageClassName` field.

Removing WIP, it's ready to be merged now.

jsafrane commented:
Added release note, please check my English.

jsafrane commented:
@k8s-bot bazel test this

msau42 commented on pkg/api/helpers.go:
```diff
@@ -557,3 +557,43 @@ func PodAnnotationsFromSysctls(sysctls []Sysctl) string {
+
+// GetPersistentVolumeClass returns StorageClassName.
+func GetPersistentVolumeClass(volume *PersistentVolume) string {
+	// Use StorageClassName attribute first
```
Don't we want beta annotation first if it's not ""?

jsafrane replied:
You're rigth, last minute edits... It's fixed now.

jsafrane committed:
Move PV/PVC annotations to PV/PVC types.

They aren't part of storage.k8s.io/v1 or v1beta1 API.
Also remove associated *GetClass functions.

msau42 commented:
This LGTM.  I would not rebase until we can recommit the storage v1 API group PR.

thockin commented:
@enisoc This PR is part of the feature family for promoting the StoragClass API.  Most of teh PRs are in or were in, and reverted for GKE issues, and are in flight again.

Do we need to go through https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md

@jsafrane @msau42

msau42 commented:
I believe this one is being replaced by PR #42128.

jsafrane commented:
sure, closing
